ui/jobs/jobs.js

```diff
@@ -3045,6 +3069,119 @@ class Bars {
     });
   }
 
+  setupDnc() {
+    const comboTimer = this.addTimerBar({
+      id: 'dnc-timers-combo',
+      fgColor: 'dnc-color-combo',
+    });
+    this.comboFuncs.push((skill) => {
+      comboTimer.duration = 0;
+      if (this.combo.isFinalSkill)
+        return;
+      if (skill)
+        comboTimer.duration = 15;
+    });
+
+    const standardStep = this.addProcBox({
+      id: 'dnc-procs-standardstep',
+      fgColor: 'dnc-color-standardstep',
+    });
+    this.abilityFuncMap[kAbility.StandardStep] = () => {
+      standardStep.duration = 0;
+      standardStep.duration = 30;
+    };
+
+    // TechnicalStep cooldown on begin dance, but effect appear when TechnicalFinish.
+    const technicalStep = this.addProcBox({
+      id: 'dnc-procs-technicalstep',
+      fgColor: 'dnc-color-technicalstep',
+    });
+    this.abilityFuncMap[kAbility.TechnicalStep] = () => {
+      technicalStep.duration = 0;
+      technicalStep.duration = 120;
+    };
+    let technicalFlag = false;
+    let elapsed = 0;
+    this.abilityFuncMap[kAbility.TechnicalFinish] = () => {
+      // Avoid mutiple call in one TechnicalFinish.
```

Comment:
I remember she said that TF is an AoE skill, so it would deal damage on multiple mobs, and produce many loglines.

Reply:
TechnicalFinish not only hit mobs, but also "hit" party member with a same ID.
If this func call more than once, `elapsed = technicalStep.elapsed` will be set to 0, makes the countdown error.

Reply:
Ah, that makes sense.  I stared at this a little bit more, and I think I understand elapsed is here to capture the difference in time between TechnicalStep and TechnicalFinish so that you can show 20 seconds for TechnicalFinish but still go back to 120 seconds for TechnicalStep from the original time TechnicalStep was pressed.  Got it!

I do have some other questions, though.

I think then it might be better to say `if (technicalFlag) return;` so that each of the extra hits doesn't create an individual timeout?

Also, like my comment below, maybe it would also help if `technicalFlag` was called something like `technicalActive` or `technicalIsActive` too, just because "flag" doesn't mean much on its own to me, and it might make it more readable.

Finally, it looks like kAbility.TechnicalFinish is really kAbility.QuadrupleTechnicalFinish.  Can you add the other technical finish ability ids, so that even if you mess up the dance it still behaves the same?

Reply:
Done. `kAbility.TechnicalFinish` seems not used in other place, I change it to `kAbility.QuadrupleTechnicalFinish`.